sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -913,10 +913,17 @@ object CollapseRepartition extends Rule[LogicalPlan] {
       case (false, true) => if (r.numPartitions >= child.numPartitions) child else r
       case _ => r.copy(child = child.child)
     }
+    case r @ Repartition(_, _, p @ Project(_, child: RepartitionOperation)) =>
```

Comment:
Is `Project` the only node to deal with? how about `Filter`/`Union`?

Reply:
This reminds me of `EliminateSorts`

Reply:
Not all `RepartitionOperation` can be removed. Sometimes repartition before joining and filtering to increase parallelism, especially before `BroadcastNestedLoopJoin`.

Reply:
`Project` can also contain expensive computation (e.g. UDF), why it's safe to remove repartition before `Project`?